src/embind/embind.js

```diff
@@ -608,53 +611,80 @@ var LibraryEmbind = {
     '$simpleReadValueFromPointer', '$throwBindingError'],
   _embind_register_std_string: function(rawType, name) {
     name = readLatin1String(name);
+    var embindStdStringUTF8Support
+#if EMBIND_STD_STRING_UTF8_SUPPORT
+    = true;
+#else
+    = false;
+#endif
     registerType(rawType, {
         name: name,
         'fromWireType': function(value) {
             var length = HEAPU32[value >> 2];
-            var a = new Array(length);
-            for (var i = 0; i < length; ++i) {
-                a[i] = String.fromCharCode(HEAPU8[value + 4 + i]);
+
+            function mapsToStdString(name) {
+                //process only std::string bindings with UTF8 support,
+                //in contrast to e.g. std::basic_string<unsigned char>
+                return name === "std::string";
             }
+
+            var str;
+            if(embindStdStringUTF8Support && mapsToStdString(name)) {
+                str = Module['UTF8ToString'](value + 4, length);
```

Comment:
What does this do if the bytes don't form a valid UTF-8 sequence? I think ideally a BindingError would be thrown. Of course, you don't want a JavaScript exception thrown through C++ though...